Plugin/Recipe system for mBuild using setuptools.

justinGilmer opened the pull request:
A working version of a plugin/recipe system for mBuild using `setuptools.entrypoints`.

Will add documentation in a few mins to walk through the basics of adding a recipe to mBuild. 

Essentially:
* You create your python module containing your recipe, `in the future, we will have a forked version of a `cookiecutter` template for creating a plugin.
* In your setup.py file add:
 ```python3
entry_points={
        'mbuild.plugins':[
            "RECIPE_NAME = path.to.module:CLASS_IN_MODULE",
       ],
```

Where `mbuild.plugins` is the name of the entry point location in the main `mbuild` project.

`RECIPE_NAME` is the name you want to give your recipe when adding it to mBuild. We recommend using the class name is possible

`path.to.module` is the path from your `setup.py` file to your python file containing your class you want to make discoverable in mBuild.
For example: in `mbuild` itself there is a recipe for creating [monolayers](https://github.com/mosdef-hub/mbuild/blob/master/mbuild/recipes/monolayer.py#L11).
Adding that to the `entry_points` so it behaves like a plugin as well:

 ```python3
entry_points={
        'mbuild.plugins':[
            "Monolayer = mbuild.recipes.monolayer:Monolayer",
       ],
```

Finally, to make your plugin discoverable in mBuild, you need to:
1. Install `mBuild` first
2. Install your project second, with the edits in your `setup.py` file. You do **not** need to do anything to mBuild.
3. see if mBuild found your project
```python3
import mbuild as mb
monolayer = mb.my_recipes.Monolayer()
```

Where `my_recipes` is an object that contains instances of all the available plugins that mBuild could find.

mattwthompson commented:
This approach seems reasonable from the user's end but it may be a doozy to test. In addition to a cookiecutter, should we have a separate repo that installs a small recipe (doesn't need to be anything big, say just a class with a name) and make sure mbuild can find it? It should be doable in a small test environment if - correct me if I'm wrong here - 

* a normal user can do it in just a few lines
* it doesn't require changes to the main codebase (a.k.a. a `conda`/`pip` install would work)
* checking its "install" can be done just by asserting the import of the plugin, maybe doing some other checks to make sure it's the right object

In this usage:
```python
import mbuild as mb
monolayer = mb.recipes.my_recipes.Monolayer()
```

why are we adding `my_recipes` inside of `recipes` vs. having the user's added recipes tacked on to `mb.recipes` itself? The way I think of it, `mBuild` ships with some recipes but the user desires to add them on to that list. If it's a matter of distinguishing between the source, I wonder if there's a cleaner way to do that that wouldn't result in going three layers deep from the main namespace?

justinGilmer commented:
> This approach seems reasonable from the user's end but it may be a doozy to test. 

Just updated the tests, realized I had not done that.


>In addition to a cookiecutter, should we have a separate repo that installs a small recipe (doesn't need to be anything big, say just a class with a name) and make sure mbuild can find it? It should be doable in a small test environment if - correct me if I'm wrong here -

It could be done either way, I think I like this was a bit more, since the testing is now checking using the `entry_point` method.

> 
> * a normal user can do it in just a few lines
> * it doesn't require changes to the main codebase (a.k.a. a `conda`/`pip` install would work)
> * checking its "install" can be done just by asserting the import of the plugin, maybe doing some other checks to make sure it's the right object

I think we could do that, create a test recipe, host it on conda, and install that for testing. Should be pretty easy to do that.


> 
> In this usage:
> 
> ```python
> import mbuild as mb
> monolayer = mb.recipes.my_recipes.Monolayer()
> ```
> 
> why are we adding `my_recipes` inside of `recipes` vs. having the user's added recipes tacked on to `mb.recipes` itself? 

The most intuitive method I could get working was to create a recipe object, called `my_recipe` and each of its attributes are the discovered entry_points plugins.

I chose `my_recipe` since it is an instance of the `Recipe` class I created. We only really want one instance of this class floating around. we could just name it`Recipe`, but it could cause import issues if we arent careful.

>The way I think of it, `mBuild` ships with some recipes but the user desires to add them on to that list. If it's a matter of distinguishing between the source, I wonder if there's a cleaner way to do that that wouldn't result in going three layers deep from the main namespace?

And I just had typed it wrong, I updated the first comment to show that. To access a recipe. It is 
`mb.my_recipes.RECIPE_NAME`

justinGilmer commented:
Ive become more and more of a fan of calling it `my_recipes` since it will be unique to each user depending on what recipes you install. And you can add them at any time, since every time you re-import mBuild, it checks your env for recipes.

justinGilmer commented:
This will break functionality for some users scripts, but they would only have to change where they incorporate one of these recipes. Which can be done with find and replace in any text editor.

justinGilmer commented:
@mikemhenry If you have any custom recipes of yours that can be converted to a discoverable recipe in mBuild using this PR, would love to see if you are able to just edit your `setup.py` and have it discoverable in mBuild.

We ran through a successful test with @rmatsum836 , but more eyes on this, the better.

mattwthompson commented:
I would still rather have it simply called `recipes`. The recipes that ship with the main repo shouldn't be in anything called `my_recipes` because they are not owned by the user. It `my_recipes` contains all recipes - those shipped by default and those added via plugins - it should just be `recipes`. I don't think a high-level import of any package be `mb.my_foo`. If it's a matter of ownership I would rather have that tracked via an attribute or something.

mattwthompson commented:
As per our discussions, I think `my_recipes` should be changed to be simply `recipes`. This leads to a name mismatch between the module implicitly named `mb.recipes` due to the name of the folder `mbuild/recipes` and the Python object `mb.recipes`. I think the best way to resolve this is move the existing "recipes" into `mbuild/lib/`.

Ties into #502

mikemhenry commented:
I might have missed it, but can you add some documentation--even if its in the readme, that it exits and link to a cookie-cutter repo that people can use that has the setup.py stuff needed?

mikemhenry commented on mbuild/examples/alkane/alkane.py:
```diff
@@ -24,7 +24,7 @@ def __init__(self, n=3, cap_front=True, cap_end=True):
         if not cap_end:
             n += 1
-        chain = mb.Polymer(CH2(), n=n-2, port_labels=('up', 'down'))
+        chain = mb.my_recipes.Polymer(CH2(), n=n-2, port_labels=('up', 'down'))
```
foo

mattwthompson replied:
I am going to resolve this comment since it was used to test permissions 😉

mattwthompson commented:
Travis has been backlogged for a day but everything passes locally. I think this does what we want - I can install Ray's porebuilder and access both its classes and mbuild's defaults

```python3
>>> import mbuild as mb
>>> print(vars(mb.recipes))
{'GraphenePore': <class 'porebuilder.porebuilder.GraphenePore'>, 'GraphenePoreSolvent': <class 'porebuilder.porebuilder.GraphenePore'>, 'Monolayer': <class 'mbuild.lib.recipes.monolayer.Monolayer'>, 'Polymer': <class 'mbuild.lib.recipes.polymer.Polymer'>, 'SilicaInterface': <class 'mbuild.lib.recipes.silica_interface.SilicaInterface'>, 'TiledCompound': <class 'mbuild.lib.recipes.tiled_compound.TiledCompound'>}

```

Still missing docs

mattwthompson commented:
@rmatsum836 can you verify you are able to import a recipe like this? i.e. make sure the second line autocompletes like it does for me


```python3
>>> import mbuild as mb
>>> mb.recipes.GraphenePoreSolvent
<class 'porebuilder.porebuilder.GraphenePoreSolvent'>
```

rmatsum836 commented:
I'm able to import `GraphenePoreSolvent` recipe with autocomplete.

mattwthompson commented:
For some reason there was a `from mbuild.recipes import *` floating around in the highest-level `__init__.py` that was not breaking anything here but made the above imports only work inside of mb. Not sure why it was there, and I find it strange that it wasn't causing problems, but removing it seems to work fine.
